cmd/reaper/types_watcher.go

```diff
@@ -0,0 +1,101 @@
+package main
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/boskos/common"
+)
+
+type ResourceTypesWatcher struct {
+	config        string
+	providedTypes []string
+	mergedTypes   []string
+	l             sync.Locker
+}
+
+func NewResourceTypesWatcher(providedTypes []string, config string) (*ResourceTypesWatcher, error) {
+	rt := ResourceTypesWatcher{
+		config:        config,
+		providedTypes: providedTypes,
+		l:             &sync.Mutex{},
+	}
+
+	if config != "" {
+		// Read and parse the types for the first time
+		// In this way 'ResourceTypesWatcher.Types()' returns the correct result
+		// even if the config is never updated
+		configResources, err := parseResourceTypesFromConfig(rTypesConfig)
+		if err != nil {
+			return nil, fmt.Errorf("parse resource types: %w", err)
+		}
+		rt.mergedTypes = mergeResourceTypes(configResources, providedTypes)
+		rt.watchConfig()
+	} else {
+		rt.mergedTypes = providedTypes
+	}
+
+	return &rt, nil
+}
+
+func (rtw *ResourceTypesWatcher) watchConfig() {
+	go func() {
+		v := viper.New()
+		v.SetConfigFile(rtw.config)
+		v.SetConfigType("yaml")
+		v.OnConfigChange(func(in fsnotify.Event) {
+			// Don't block the handler on reading and parsing the new config
+			go func() {
+				logrus.Info("a new boskos config is available")
+				newResourceTypes, err := parseResourceTypesFromConfig(rtw.config)
+				if err != nil {
+					logrus.WithError(err).Warn("failed to parse types from a new config")
+					return
+				}
+				newMergedTypes := mergeResourceTypes(newResourceTypes, rtw.providedTypes)
+
+				// Keep the critical region as small as possible
+				func() {
+					rtw.l.Lock()
+					defer rtw.l.Unlock()
+					rtw.mergedTypes = newMergedTypes
+				}()
+
+				logrus.Infof("new resources: %+v", rtw.mergedTypes)
+			}()
+		})
+		v.WatchConfig()
+	}()
+}
+
+func (rtw *ResourceTypesWatcher) Types() []string {
+	// There is room for a small optimization here:
```

Comment:
If your constructor returns an interface (or even just the getter func) you can just have a simple closure that returns the list when no file is given